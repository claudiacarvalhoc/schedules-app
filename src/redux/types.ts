export const FETCH_SCHEDULE_SUCCESS = 'FETCH_SCHEDULE_SUCCESS';
export const FETCH_SCHEDULE_FAILURE = 'FETCH_SCHEDULE_FAILURE';

export const FETCH_SCHEDULE_LOGS_SUCCESS = 'FETCH_SCHEDULE_LOGS_SUCCESS';
export const FETCH_SCHEDULE_LOGS_FAILURE = 'FETCH_SCHEDULE_LOGS_FAILURE';

export const REQUEST_RETIRE_SCHEDULE = 'REQUEST_RETIRE_SCHEDULE';
export const REQUEST_UNRETIRE_SCHEDULE = 'REQUEST_UNRETIRE_SCHEDULE';
export const UPDATE_SELECTED_SCHEDULE = 'UPDATE_SELECTED_SCHEDULE';