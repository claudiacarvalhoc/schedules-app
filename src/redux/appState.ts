import { DefaultRootState } from "react-redux";

/**
 * The App Status
 */
export interface AppState extends DefaultRootState {
    status: AppStatusSectionsState;
    selectedScheduleId?: number;
    schedules: ScheduleState[];
    scheduleLogs: ScheduleLogState[];
    texts: {
        header: HeaderTextStatus;
        schedules: ScheduleTextState;
    }
}

/**
 * The status to fetch data per section
 */
export interface AppStatusSectionsState {
    schedules: StatusState;
    scheduleLogs: StatusState;
}

/**
 * The Boot Status is defined for options: Loading / Success / Failure.
 */
export enum StatusState {
    Loading,
    Success,
    Failure
}

/**
 * The Schedule interface.
 */
export interface ScheduleState {
    id: number;
    avatar: AvatarState;
    name: string;
    description: string;
    isRetired: boolean;
    tasksCount: number;
    startPoint: Date;
    endPoint: Date;
    dayOfWeek: number;
    dayOfMonth: number;
    startDate: Date;
    endDate: Date;
    schedules?: IntervalTypeState;
    timePeriod?: number;
}

/**
 * The Avatar Identification
 */
export interface AvatarState {
    letter: string;
    color: string;
}

/**
 * The Schedule Interval Types
 */
export enum IntervalTypeState {
    Once,
    Year,
    Week,
    Hour,
    Minute,
    Second
}

/**
 * The Schedule Logs interface.
 */
export interface ScheduleLogState {
    id: number;
    startTime: Date;
    endTime: Date;
    status: LogStatusState;
    serverName: string;
    scheduleId: number;
}

/**
 * The Schedule Logs Status
 */
export enum LogStatusState {
    Terminated,
    Pending,
    Completed,
    Exception,
    Running
}

/**
 * Texts used on Header
 * Note: This allow us to add localisation feature easily
 */
export interface HeaderTextStatus {
    titleText: string;
}

/**
 * Texts used on Schedule
 * Note: This allow us to add localisation feature easily
 */
export interface ScheduleTextState {
    buttonRetireText: string;
    buttonUnretireText: string;
    errorMessageText: string;
    buttonRetryText: string;
    emptyMessageText: string;
}
