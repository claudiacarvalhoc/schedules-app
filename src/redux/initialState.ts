import { AppState, AvatarState, Status } from "./appState";

/**
 * This is the initial state of redux.
 */
export const appInitialState: AppState = {
    status: {
        schedules: Status.Loading,
        scheduleLogs: Status.Loading,
    },
    selectedScheduleId: undefined,
    schedules: [],
    scheduleLogs: [],
    texts: {
        header: {
            titleText: 'Schedules',
        },
        schedules: {
            buttonRetireText: 'Retire',
            buttonUnretireText: 'Unretire',
            errorMessageText: 'Something went wrong ...',
            buttonRetryText: 'Retry',
            emptyMessageText: 'No schedules available.',
        }
    }
};

export const avatarInitialState: AvatarState[] = [
    {
      letter: 'N',
      color: 'orange',
    },
    {
      letter: 'A',
      color: 'purple',
    },
    {
      letter: 'T',
      color: 'orange',
    },
    {
      letter: 'J',
      color: 'purple',
    }
];