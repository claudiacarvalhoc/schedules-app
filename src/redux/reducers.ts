import { Reducer, combineReducers } from 'redux';
import { AppState } from './appState';
import { appReducer } from './app/reducer';

export const rootReducer = combineReducers({
    app: appReducer as Reducer<AppState>,
});

export type RootState = ReturnType<typeof rootReducer>;
export type DefaultRootState = ReturnType<typeof rootReducer>;