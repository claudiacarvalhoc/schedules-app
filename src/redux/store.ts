import { AnyAction } from 'redux';
import { AppState } from './appState';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { applyMiddleware, compose, createStore } from 'redux';
import { createLogger } from 'redux-logger';
import { rootReducer } from './reducers';
import thunk from 'redux-thunk';

export type AppThunk<R> = ThunkAction<R, AppState, unknown, AnyAction>;
export type AppDispatch = ThunkDispatch<AppState, unknown, AnyAction>;


export const configureStore = (initialState: AppState) => {
    const logger = createLogger();
    const composeEnhancers = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
    const middleware = [thunk, logger];

    const store = createStore(
        rootReducer,
        {
            app: initialState,
        },
        composeEnhancers(applyMiddleware(...middleware)),
    );

    return store;
};

export default configureStore;
