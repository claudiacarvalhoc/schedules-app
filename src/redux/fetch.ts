import { AppDispatch } from './store';

export const getSchedules = (): ((dispatch: AppDispatch) => Promise<void>) => {
    return dispatch => {
        // TODO : ....
        return Promise.resolve();
    };
};

export const getScheduleLogs = (): ((dispatch: AppDispatch) => Promise<void>) => {
    return dispatch => {
        // TODO : ....
        return Promise.resolve();
    };
};
