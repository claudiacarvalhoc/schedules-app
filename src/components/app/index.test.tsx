import React from 'react';
import App from './index';
import Enzyme, { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { appInitialState } from '../../redux/initialState';
import thunk from 'redux-thunk';
import { StatusState } from '../../redux/appState';

Enzyme.configure({ adapter: new Adapter() });

const getMountWrapper = (status: StatusState = undefined) => {
    const store = configureMockStore([thunk])({
      app: {
        ...appInitialState,
        status: {
          ...appInitialState.status,
          schedules: status ?? appInitialState.status.schedules,
          scheduleLogs: status ?? appInitialState.status.scheduleLogs,
        },
      }
    });
    return mount(
        <Provider store={store}>
            <App />
        </Provider>
    );
  };

describe('<Header />', () => {
    it('render correctly when data is being fetched', () => {
        const component = getMountWrapper();
        expect(toJson(component)).toMatchSnapshot();
        expect(component.find('[data-test="app_show_loading"]').exists()).toBeTruthy();
    });
    it('render correctly when data is fetched', () => {
      const component = getMountWrapper(StatusState.Success);
      expect(toJson(component)).toMatchSnapshot();
      expect(component.find('[data-test="app_load_schedules"]').exists()).toBeTruthy();
  });
});
