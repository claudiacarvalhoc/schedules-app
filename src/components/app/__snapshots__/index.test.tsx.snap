// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(App)>
    <App
      bootApp={[Function]}
    >
      <div
        className="container"
      >
        <Connect(Header)>
          <Header
            dispatch={[Function]}
            title="Schedules"
          >
            <div
              className="container"
            >
              <WithStyles(ForwardRef(Typography))
                className="title"
                gutterBottom={true}
                variant="h5"
              >
                <ForwardRef(Typography)
                  className="title"
                  classes={
                    Object {
                      "alignCenter": "MuiTypography-alignCenter",
                      "alignJustify": "MuiTypography-alignJustify",
                      "alignLeft": "MuiTypography-alignLeft",
                      "alignRight": "MuiTypography-alignRight",
                      "body1": "MuiTypography-body1",
                      "body2": "MuiTypography-body2",
                      "button": "MuiTypography-button",
                      "caption": "MuiTypography-caption",
                      "colorError": "MuiTypography-colorError",
                      "colorInherit": "MuiTypography-colorInherit",
                      "colorPrimary": "MuiTypography-colorPrimary",
                      "colorSecondary": "MuiTypography-colorSecondary",
                      "colorTextPrimary": "MuiTypography-colorTextPrimary",
                      "colorTextSecondary": "MuiTypography-colorTextSecondary",
                      "displayBlock": "MuiTypography-displayBlock",
                      "displayInline": "MuiTypography-displayInline",
                      "gutterBottom": "MuiTypography-gutterBottom",
                      "h1": "MuiTypography-h1",
                      "h2": "MuiTypography-h2",
                      "h3": "MuiTypography-h3",
                      "h4": "MuiTypography-h4",
                      "h5": "MuiTypography-h5",
                      "h6": "MuiTypography-h6",
                      "noWrap": "MuiTypography-noWrap",
                      "overline": "MuiTypography-overline",
                      "paragraph": "MuiTypography-paragraph",
                      "root": "MuiTypography-root",
                      "srOnly": "MuiTypography-srOnly",
                      "subtitle1": "MuiTypography-subtitle1",
                      "subtitle2": "MuiTypography-subtitle2",
                    }
                  }
                  gutterBottom={true}
                  variant="h5"
                >
                  <h5
                    className="MuiTypography-root title MuiTypography-h5 MuiTypography-gutterBottom"
                  >
                    Schedules
                  </h5>
                </ForwardRef(Typography)>
              </WithStyles(ForwardRef(Typography))>
              <ForwardRef(DehazeIcon)>
                <WithStyles(ForwardRef(SvgIcon))>
                  <ForwardRef(SvgIcon)
                    classes={
                      Object {
                        "colorAction": "MuiSvgIcon-colorAction",
                        "colorDisabled": "MuiSvgIcon-colorDisabled",
                        "colorError": "MuiSvgIcon-colorError",
                        "colorPrimary": "MuiSvgIcon-colorPrimary",
                        "colorSecondary": "MuiSvgIcon-colorSecondary",
                        "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                        "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                        "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                        "root": "MuiSvgIcon-root",
                      }
                    }
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M2 15.5v2h20v-2H2zm0-5v2h20v-2H2zm0-5v2h20v-2H2z"
                      />
                    </svg>
                  </ForwardRef(SvgIcon)>
                </WithStyles(ForwardRef(SvgIcon))>
              </ForwardRef(DehazeIcon)>
            </div>
          </Header>
        </Connect(Header)>
        <div
          className="content"
        >
          <Connect(Schedules)>
            <Schedules
              fetchSchedules={[Function]}
              schedules={Array []}
              status={0}
              texts={
                Object {
                  "buttonRetireText": "Retire",
                  "buttonRetryText": "Retry",
                  "buttonUnretireText": "Unretire",
                  "emptyMessageText": "No schedules available.",
                  "errorMessageText": "Something went wrong ...",
                }
              }
            >
              <div
                className="container"
              >
                <div
                  className="loading"
                >
                  <Loading
                    color="#000000"
                    delay={0}
                    height={50}
                    type="spin"
                    width={50}
                  >
                    <div
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\">
  <path opacity=\\".25\\" d=\\"M16 0 A16 16 0 0 0 16 32 A16 16 0 0 0 16 0 M16 4 A12 12 0 0 1 16 28 A12 12 0 0 1 16 4\\"/>
  <path d=\\"M16 0 A16 16 0 0 1 32 16 L28 16 A12 12 0 0 0 16 4z\\">
    <animateTransform attributeName=\\"transform\\" type=\\"rotate\\" from=\\"0 16 16\\" to=\\"360 16 16\\" dur=\\"0.8s\\" repeatCount=\\"indefinite\\" />
  </path>
</svg>
",
                        }
                      }
                      style={
                        Object {
                          "fill": "#000000",
                          "height": 50,
                          "width": 50,
                        }
                      }
                    />
                  </Loading>
                </div>
              </div>
            </Schedules>
          </Connect(Schedules)>
          <Connect(ScheduleLogs)>
            <ScheduleLogs
              dispatch={[Function]}
              logs={Array []}
              status={0}
            >
              <div>
                <div>
                  <Loading
                    color="#000000"
                    delay={0}
                    height={50}
                    type="spin"
                    width={50}
                  >
                    <div
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\">
  <path opacity=\\".25\\" d=\\"M16 0 A16 16 0 0 0 16 32 A16 16 0 0 0 16 0 M16 4 A12 12 0 0 1 16 28 A12 12 0 0 1 16 4\\"/>
  <path d=\\"M16 0 A16 16 0 0 1 32 16 L28 16 A12 12 0 0 0 16 4z\\">
    <animateTransform attributeName=\\"transform\\" type=\\"rotate\\" from=\\"0 16 16\\" to=\\"360 16 16\\" dur=\\"0.8s\\" repeatCount=\\"indefinite\\" />
  </path>
</svg>
",
                        }
                      }
                      style={
                        Object {
                          "fill": "#000000",
                          "height": 50,
                          "width": 50,
                        }
                      }
                    />
                  </Loading>
                </div>
              </div>
            </ScheduleLogs>
          </Connect(ScheduleLogs)>
        </div>
      </div>
    </App>
  </Connect(App)>
</Provider>
`;
